sound/soc/sof/sof-client-dma-trace.c

```diff
@@ -0,0 +1,690 @@
+			goto trace_release;
+		}
+	}
+
+	priv->dtrace_is_enabled = true;
+
+	return 0;
+
+trace_release:
+	if (ops)
+		ops->release(cdev);
+
+	return ret;
+}
+
+static void sof_dtrace_fw_state(struct sof_client_dev *cdev,
+				enum snd_sof_fw_state state)
+{
+	struct sof_dtrace_priv *priv = cdev->data;
+
+	if (priv->dtrace_is_enabled && state == SOF_FW_CRASHED) {
+		priv->dtrace_error = true;
+		wake_up(&priv->dtrace_sleep);
+	}
+}
+
+static int sof_dtrace_client_probe(struct auxiliary_device *auxdev,
+				   const struct auxiliary_device_id *id)
+{
+	struct sof_client_dev *cdev = auxiliary_dev_to_sof_client_dev(auxdev);
+	enum sof_dtrace_platform dtrace_platform = id->driver_data;
+	struct dentry *dfsroot = sof_client_get_debugfs_root(cdev);
+	struct device *dma_dev =  sof_client_get_dma_dev(cdev);
+	struct sof_dma_trace_host_ops *ops;
+	struct device *dev = &auxdev->dev;
+	struct sof_dtrace_priv *priv;
+	int ret;
+
+	/* do not set up the dma-trace support if it is not enabled */
+	if (!sof_dtrace_enabled)
+		return -ENXIO;
+
+	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
+	if (!priv)
+		return -ENOMEM;
+
+	ops = dev->platform_data;
+
+	if (dtrace_platform == SOF_DTRACE_PLATFORM_HOST_ASSISTED && (!ops ||
+	    !ops->init || !ops->release || !ops->start || !ops->stop)) {
+		dev_err(dev, "missing platform callback(s) for host assisted dma-trace\n");
+		return -ENODEV;
+	}
+
+	/*
+	 * dma-trace is power managed via auxdev suspend/resume callbacks by
+	 * SOF core
+	 */
+	pm_runtime_no_callbacks(dev);
+
+	priv->host_ops = ops;
+	priv->dev = dev;
+	cdev->data = priv;
+
+	/* allocate trace page table buffer */
+	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, dma_dev, PAGE_SIZE,
+				  &priv->dmatp);
+	if (ret < 0) {
+		dev_err(dev, "can't alloc page table for trace %d\n", ret);
+		return ret;
+	}
+
+	/* allocate trace data buffer */
+	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV_SG, dma_dev,
+				  SOF_DTRACE_BUF_SIZE, &priv->dmatb);
+	if (ret < 0) {
+		dev_err(dev, "can't alloc buffer for trace %d\n", ret);
+		goto page_err;
+	}
+
+	/* create compressed page table for audio firmware */
+	ret = snd_sof_create_page_table(dma_dev, &priv->dmatb, priv->dmatp.area,
+					priv->dmatb.bytes);
+	if (ret < 0)
+		goto table_err;
+
+	priv->dtrace_pages = ret;
+	dev_dbg(dev, "dtrace_pages: %d\n", priv->dtrace_pages);
+
+	priv->dfs_trace = debugfs_create_file("trace", 0444, dfsroot, cdev,
+					      &sof_dtrace_trace_fops);
+	priv->dfs_filter = debugfs_create_file("filter", 0200, dfsroot, cdev,
+					       &sof_dtrace_filter_fops);
+
+	init_waitqueue_head(&priv->dtrace_sleep);
+
+	ret = sof_client_register_ipc_rx_handler(cdev, SOF_IPC_GLB_TRACE_MSG,
+						 snd_sof_dtrace_update_pos);
+	if (ret)
+		goto register_rx_err;
+
+	ret = sof_client_register_fw_state_handler(cdev, sof_dtrace_fw_state);
+	if (ret)
+		goto register_fw_state_err;
+
+	ret = sof_dtrace_init_ipc(cdev);
+	if (ret < 0) {
+		sof_client_unregister_ipc_rx_handler(cdev, SOF_IPC_GLB_TRACE_MSG);
```

Comment:
looks like you already have this under the `ipc_err` label?

Reply:
Oh, you are right!